Kadai3 1 mizushima

MizushimaToshihiko opened the pull request:
課題3-1を作成致しました。もしよろしければレビューいただけますと幸いです。

## 課題3-1 タイピングゲームを作ろう
- 標準出力に英単語を出す（出すものは自由）
- 標準入力から1行受け取る
- 制限時間内に何問解けたか表示する


### コマンドラインオプション

 | オプション | 説明 | デフォルト |
 | --- | --- | --- |
 | -limit | ゲーム全体の制限時間を秒数で設定する | 20 |


### インストール方法
```bash
go get github.com/MizushimaToshihiko/gopherdojo-studyroom/kadai3-1/Mizushima
```

### 使い方
1. バイナリビルド（実行ファイル作成）
```bash
$ make
```
2. ディレクトリを指定して実行
```bash
$ ./bin/typing [option]
```


### テストの方法
- バイナリビルド & テスト
```bash
$ make test
```
- テスト後の処理(掃除)
```bash
$ make clean
```

MizushimaToshihiko committed:
課題2を提出
課題1のテストを追加

MizushimaToshihiko committed:
mainのテストを追加
その他
main.go テストの追加により仕様変更
picconvert/picconvert.go godocでの見栄えが少しでも良くなるようにコメント追加
picconvert_test.go TestGlobでのテストケース追加、testdataをコピーしないように変更
Makefile 各コマンドをgoのものに変更
等々

MizushimaToshihiko committed:
コマンドラインオプションの部分を訂正

MizushimaToshihiko committed:
使い方の2を訂正

MizushimaToshihiko committed:
go fmt適用

MizushimaToshihiko committed:
README.md それぞれのカバレッジが分かりやすくなるように修正

MizushimaToshihiko committed:
kadai3-1 first commit

MizushimaToshihiko committed:
とりあえずタイピングと正解判定できるように

MizushimaToshihiko committed:
時間制限とgo.mod追加

MizushimaToshihiko committed:
単語をCSVから読み込むように変更

MizushimaToshihiko committed:
モジュール分割

MizushimaToshihiko committed:
gigitignore,Makefile,typingのテストを追加,テスト追加に伴いtyping/typing.goとmain.go変更

MizushimaToshihiko committed:
README修正

MizushimaToshihiko committed:
go fmt適用

MizushimaToshihiko committed:
READMEにディレクトリ構成を追加

MizushimaToshihiko committed:
課題2のコミットが残っていたので削除

sryoya commented on kadai3-1/Mizushima/main.go:
```diff
@@ -0,0 +1,23 @@
+)
+
+func main() {
+	wordsPath := "./gamedata/words.csv"
```
やや発展的な内容なのでFYIですが、Go1.16から追加された`go:embed` を使うとファイルの中身を直に埋め込むこともできます

https://future-architect.github.io/articles/20210208/

もちろん、このやり方だと実際に埋め込む値が一緒にビルドされてしまうので、そこをビルドなしで差し替えられるようにするなら、現状の実行時に取ってくる方法の方がベターですが。

sryoya commented on kadai3-1/Mizushima/typing/typing_test.go:
```diff
@@ -0,0 +1,68 @@
+			ans: []string{
+				"America",
+				"American",
+				"typo",
```
私ならこうするという話ですが、異常値とか結果を左右する値は、その値のところにコメントをつけるか、ぱっとみでわかる値(もっと長いとか絵文字を入れるとか)にすると、あとでテストの値をいじりたくなったときにもレビューするときにもよかったりします

sryoya commented on kadai3-1/Mizushima/typing/typing_test.go:
```diff
@@ -0,0 +1,68 @@
+	t.Helper()
+
+	cases := []struct {
+		name     string
```
これもマストではないですが、structのsliceからmapにして、nameをmapにkeyにするとわかりやすかったりします
https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

sryoya commented on kadai3-1/Mizushima/typing/typing.go:
```diff
@@ -0,0 +1,137 @@
+
+	for {
+		row, err = csvFile.Read()
+		if err != nil {
```
読む行がない場合に出る想定内のエラー(`io.EOF`)と、それ以外のエラーかのチェックをした方が良いと思います。
https://golang.org/pkg/encoding/csv/#Reader.Read
前者の場合はbreakでよくて、それ以外の場合は異常系なのでエラーを返した方が良いかと

sryoya commented on kadai3-1/Mizushima/typing/typing.go:
```diff
@@ -0,0 +1,137 @@
+	limit := time.After(t)
+
+	rand.Seed(time.Now().UnixNano())
+	var indices []int
```
重たい処理でなくて片方がテスト限定であれば、どちらかの値をこの時点で代入してしまうことで、下のelseを省略できて見通しがよくなるかと
こういうイメージです

```.go
indices := rand.Perm(len(words))
if isTest {
  indices = incSlice(len(words))
}

```

sryoya commented on kadai3-1/Mizushima/typing/typing.go:
```diff
@@ -0,0 +1,137 @@
+		case ans := <-ch:
+			// fmt.Printf("ans: %s\n", ans)
+			// fmt.Printf("word: %s\n", word)
+			if ans == word {
+				if !isTest {
+					_, err = fmt.Fprintln(w, "> しぇえか～い")
+					if err != nil {
+						return -1, err
+					}
+				}
+				score++
+			} else {
+				if !isTest {
+					_, err = fmt.Fprintln(w, "> ぶっぶー")
+					if err != nil {
+						return -1, err
+					}
+				}
+			}
```
ここもどちらかのケースでearly returnすれば、ネストが減らせますね

sryoya reviewed (commented):
こんにちは、何点かコメントしてみました 👋

MizushimaToshihiko committed:
レビュー後訂正
main.goでwords.csvをgo:embedで読み込み
typing/typing_test.goのTestGameでテストケースのOne typoのtypoを絵文字に変更
typing/typing.goのGame関数で17行目のifを削除し一旦代入した後でTestならテスト用のindexが格納されたスライスを代入し直すように変更

MizushimaToshihiko committed:
コメント追加および削除

MizushimaToshihiko committed:
typing_test.goの変更
Test関数を追加
- TestGame_TimelimitOver
- Test_input
- Test_incSlice
TestGame関数の名称変更=>TestGame_WithinTheTimelimit

MizushimaToshihiko committed:
typing.goの58行目からのifネストを修正他

MizushimaToshihiko commented:
後続の方が増えてきたのでCloseします。
